### ISSUE TITLE
`unnecessary_operation` does not treat bounds checking as a side effect

### ISSUE THREAD

**Issue by jonas-schievink** [rust-lang/rust-clippy#7408]
Lint name: `unnecessary_operation`

I tried this code:

```rust
struct ZST;

impl ZST {
    fn doit() {
        [()][core::mem::size_of::<Self>()];
    }
}
```

I expected to see this happen: the operation panics if `ZST` is not zero-sized, so the array index operation needs to be preserved.

Instead, this happened:

```
warning: statement can be reduced
 --> src/main.rs:5:9
  |
5 |         [()][core::mem::size_of::<Self>()];
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `[()];core::mem::size_of::<Self>();`
  |
  = note: `#[warn(clippy::unnecessary_operation)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_operation
```

The suggested replacement will never panic, so it does not preserve semantics.

### Meta

- `cargo clippy -V`: clippy 0.1.53 (53cb7b0 2021-06-17)
- `rustc -Vv`:
```
rustc 1.53.0 (53cb7b09b 2021-06-17)
binary: rustc
commit-hash: 53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b
commit-date: 2021-06-17
host: x86_64-unknown-linux-gnu
release: 1.53.0
LLVM version: 12.0.1
```
also happens on nightly on the playground

**Comment by y21** [comment]
This was fixed in #7453. The lint still emits a warning here, but now suggests using `assert!` instead of removing the indexing entirely.
